network/src/noise/handshake.rs

```diff
@@ -179,44 +187,54 @@ impl NoiseUpgrader {
     /// Perform an outbound protocol upgrade on this connection.
     ///
     /// This runs the "client" side of the Noise IK handshake to establish a
-    /// secure Noise stream and exchange static public keys. In mutual auth
-    /// scenarios, we will also include an anti replay attack counter in the
+    /// secure Noise stream and send its static public key to the server.
+    /// In mutual auth scenarios, we will also include an anti replay attack counter in the
     /// Noise handshake payload. Currently this counter is always a millisecond-
     /// granularity unix epoch timestamp.
-    pub async fn upgrade_outbound<TSocket>(
+    pub async fn upgrade_outbound_with_noise<TSocket>(
         &self,
         mut socket: TSocket,
         remote_public_key: x25519::PublicKey,
     ) -> io::Result<NoiseStream<TSocket>>
     where
         TSocket: AsyncRead + AsyncWrite + Unpin,
     {
-        // send a payload of the current timestamp (in milliseconds)
+        // craft prologue = self_peer_id | expected_public_key
+        let mut prologue = [0; PeerId::LENGTH + x25519::PUBLIC_KEY_SIZE];
+        prologue[..PeerId::LENGTH].copy_from_slice(self.self_peer_id.as_ref());
+        prologue[PeerId::LENGTH..].copy_from_slice(remote_public_key.as_slice());
+
+        // craft 8-byte payload as current timestamp (in milliseconds)
         let payload = {
             let now: u64 = time::SystemTime::now()
                 .duration_since(time::UNIX_EPOCH)
                 .expect("system clock should work")
-                .as_millis() as u64;
+                .as_millis() as u64; // (PAYLOAD_SIZE)
+
             // e.g. [157, 126, 253, 97, 114, 1, 0, 0]
             let now = now.to_le_bytes().to_vec();
             Some(now)
         };
 
-        // create first handshake message  (-> e, es, s, ss)
+        // buffer for the first message
+        let mut first_message = [0u8; PeerId::LENGTH
+            + x25519::PUBLIC_KEY_SIZE
+            + noise::handshake_init_msg_len(PAYLOAD_SIZE)];
+
+        // craft first handshake message  (-> e, es, s, ss)
         let mut rng = rand::rngs::OsRng;
-        let mut first_message = [0u8; noise::handshake_init_msg_len(PAYLOAD_SIZE)];
         let initiator_state = self
             .noise_config
             .initiate_connection(
                 &mut rng,
-                &[],
+                &prologue,
                 remote_public_key,
                 payload.as_ref().map(|x| &x[..]),
                 &mut first_message,
             )
```

Comment:
i think something like this should work

```rust
const PROLOGUE_SIZE = PeerId::LENGTH + x25519::PUBLIC_KEY_SIZE;
const CLIENT_MESSAGE_SIZE = PROLOGUE_SIZE + noise::handshake_init_msg_len(PAYLOAD_SIZE);

// ..

let mut client_msg = [0u8; CLIENT_MESSAGE_SIZE];
client_msg[..PeerId::LENGTH].copy_from_slice(self.self_peer_id.as_ref());
client_msg[PeerId::LENGTH..PROLOGUE_SIZE].copy_from_slice(remote_public_key.as_slice());

let (prologue_msg, client_noise_msg) = client_msg.split_at_mut(PROLOGUE_SIZE);

let payload = {
    let now: u64 = time::SystemTime::now()
        .duration_since(time::UNIX_EPOCH)
        .expect("system clock should work")
        .as_millis() as u64; // (PAYLOAD_SIZE)

    // e.g. [157, 126, 253, 97, 114, 1, 0, 0]
    now.to_le_bytes()
};

let initiator_state = self
    .noise_config
    .initiate_connection(
        &mut rng,
        &prologue,
        remote_public_key,
        Some(&payload[..]),
        &mut client_noise_msg,
    )
// ..

socket.write_all(&client_msg).await?;
```

Reply:
actually are you sure we needed one write_all since we only flushed at the end?

Reply:
alright done!

Reply:
1. tcp flush is a no-op anyway : p (https://docs.rs/tokio/0.2.21/src/tokio/net/tcp/stream.rs.html#867-870)

2. like i said, we have TCP_NODELAY turned on. without TCP_NODELAY, the socket will wait for enough data to buffer before sending (according to [Nagle's algorithm](https://en.wikipedia.org/wiki/Nagle's_algorithm)). with TCP_NODELAY it'll just send the write immediately